/********************************************************************************
* Copyright (c) 2022 Contributors to the Eclipse Foundation
*
* See the NOTICE file(s) distributed with this work for additional
* information regarding copyright ownership.
*
* This program and the accompanying materials are made available under the
* terms of the Apache License 2.0 which is available at
* http://www.apache.org/licenses/LICENSE-2.0
*
* SPDX-License-Identifier: Apache-2.0
********************************************************************************/

syntax = "proto3";

package sdv.edge.comfort.trunk.v1;

/**
 * @brief Example trunk service for controlling the trunks of the vehicle body.
 *        This definition is designed here according to the draft of the comfort seats
 *        service definition of the COVESA Vehicle Service Catalog (VSC) (see
 *        https://github.com/COVESA/vehicle_service_catalog) as a definition of an
 *        trunk service is currently missing in VSC.
 */
service Trunk {
    /** Set the desired lock state of the addressed trunk(s)
     *  Setting the lock state is possible independently of the open/closed state
     *  of the respective trunk lid; in particular it is possible to set the state
     *  to locked during the lid is open.
     *
     *  Returns gRPC status codes:
     *   * OK - The lock state was sucessfully set
     *   * NOT_FOUND - The specified trunk instance is not available in this vehicle
     *   * UNIMPLEMENTED - Setting the lock state is not supported by this vehicle
     *   * INTERNAL - A service internal error happened - see error message for details
    */
    rpc SetLockState(SetLockStateRequest) returns (SetLockStateReply);

    /** Request to open the addressed trunk(s)
     *  Depending on the underlying vehicle hardware, this
     *   - could be not supported at all,
     *   - the request triggers the trunk lid to just "spring open", or
     *   - the lid is moved to fully open state by a spring or a motor.
     *  The request to open (if supported) does open the trunk lid, but does not
     *  change the lock state of the trunk.
     *
     *  Returns gRPC status codes:
     *   * OK - The operation to open the trunk was successfully triggered.
     *   * NOT_FOUND - The specified trunk instance is not available in this vehicle
     *   * UNIMPLEMENTED - Opening the trunk is not supported by this vehicle
     *   * INTERNAL - A service internal error happened - see error message for details
    */
    rpc Open(OpenRequest) returns (OpenReply);

    /** Request to close the addressed trunk(s)
     *  Depending on the underlying vehicle hardware, this
     *   - could be not supported at all, or
     *   - the lid will be moved to closed state, e.g. by a motor.
     *  This request (if supported) does close the trunk lid, but does not
     *  change the lock state of the trunk.
     *
     *  Returns gRPC status codes:
     *   * OK - The operation to close the trunk(s) was successfully triggered.
     *   * NOT_FOUND - The specified trunk instance is not available in this vehicle
     *   * UNIMPLEMENTED - Closing the trunk is not supported by this vehicle
     *   * INTERNAL - A service internal error happened - see error message for details
    */
    rpc Close(CloseRequest) returns (CloseReply);
}

message SetLockStateRequest {
    TrunkInstance instance = 1;
    LockState state = 2;
}

message SetLockStateReply {}

message OpenRequest {
    TrunkInstance instance = 1;
}

message OpenReply {}

message CloseRequest {
    TrunkInstance instance = 1;
}

message CloseReply {}

enum TrunkInstance {
    ALL = 0;
    FRONT = 1;
    REAR = 2;
}

enum LockState {
    LOCKED = 0;
    UNLOCKED = 1;
}